robotPositioner/robotPositioner.h

```diff
@@ -0,0 +1,382 @@
+        T det = a*d - b*c;
+
+        // if the determinant is not zero, there is a solution
+        if (det != 0) {
+            Vl = (e*d - b*f)/det;
+            Vr = (a*f - e*c)/det;
+        }
+        // drive robot 
+        m_bot->tank(Vl, Vr);
+    }
+
+    // captures a frame
+    void captureCameraFrame() {
+        const cv::Size unwrapSize(180,50);
+        cv::Mat output(m_cam->getSuperPixelSize(), CV_8UC1);
+        cv::Mat unwrapped(unwrapSize, CV_8UC1);
+
+        if (m_cam->captureSuperPixelGreyscale(output)) {
+            m_unwrapper.unwrap(output, unwrapped);
+            char fileName[255];
+            sprintf(fileName, "image_%u.png", m_imageNumber++);
+            cv::imwrite(fileName, unwrapped);
+            std::cout << " image " << m_imageNumber << " captured " << std::endl;
+        }
+    }
+
+//-----------------PUBLIC API---------------------------------------------------------------------
+public:
+
+    // constructor
+    robotPositioner(
+        unsigned int vicon_udp_client_port,
+        std::string  vicon_capture_control_ip_address,
+        std::string  vicon_capture_control_executable_path,
+        unsigned int capture_control_port,
+        double       robot_r,
+        double       robot_D,
+        double       allowed_heading_error,
+        double       stopping_distance,
+        std::string  video_device,
+        double       k1,
+        double       k2,
+        double       alpha,
+        double       beta,
+        double       max_velocity
+        ) : m_vicon_udp_client_port(vicon_udp_client_port),
+            m_vicon_capture_control_ip_address(vicon_capture_control_ip_address),
+            m_vicon_capture_control_executable_path(vicon_capture_control_executable_path),
+            m_capture_control_port(capture_control_port),
+            m_robot_r(robot_r),
+            m_robot_D(robot_D),
+            m_allowed_heading_error(allowed_heading_error),
+            m_stopping_distance(stopping_distance),
+            m_video_device(video_device),
+            m_k1(k1),
+            m_k2(k2),
+            m_alpha(alpha),
+            m_beta(beta),
+            m_max_velocity(max_velocity)
+        {
+
+        m_imageNumber = 0;
+        m_bot = new BoBRobotics::Robots::Norbot(); // init robot
+        setupVicon();                              // init vicon system
+
+        // setup 360 camera if exists
+        try {
+            setupCamera360();
+            std::cout << "camera 360 finished setting up" << std::endl;
+        } catch(...) {
+            std::cout << "camera cannot be initialised, maybe it does not exists..." << std::endl;
+            m_hasCamera = false;
+        }
+    }
+
+    //destructor
+    ~robotPositioner() {
+
+        if(!m_viconCaptureControl->stopRecording("test1")) {
+            std::cout << "failed to stop recording" << std::endl;
+        }
+
+        delete m_vicon;
+        delete m_viconCaptureControl;
+        delete m_bot;
+
```

Comment:
As mentioned before, it would be best to move these objects outside the scope of the class, but aside from that you also don't really want to use ``new`` and ``delete`` in modern C++. For the most part you can avoid using pointers at all and just pass by reference, but if you have to it's much better to use some kind of smart pointer.

Reply:
I should probably clarify that I'm not saying regular pointers are always bad! It's just ``new`` and ``delete`` you want to avoid. It's fine to pass a regular pointer so long as some other scope "owns" it and will clean it up (e.g. because it's using a ``unique_ptr`` or something)